### PR TITLE
PE-7059: Buy records with various funding sources

### PR DESCRIPTION
**Pull request by arielmelendez**

### CODE REVIEW

File: src/main.lua [hunk 1]
```diff
@@ -993,40 +1023,64 @@ addEventingHandler(ActionMap.TokenCost, utils.hasMatchingTag("Action", ActionMap
...
+		return
 	end
+
+	ao.send({
+		Target = msg.From,
+		Tags = { Action = "Token-Cost-Notice", ["Token-Cost"] = tostring(tokenCost) },
+		Data = fundFrom and json.encode({
+				tokenCost = tokenCost,
+				fundingPlan = fundingPlan,
+			})
```
Comment (dtfiedler) [inline src/main.lua:1080]: mentioned in https://github.com/ar-io/ar-io-network-process/pull/172 - let's also include `discounts` as the result from `getTokenCost`

File: src/gar.lua [hunk 2]
```diff
@@ -1191,4 +1196,259 @@ function gar.delegateAllowedToStake(delegateAddress, gateway)
...
+	return fundingPlan
+end
+
+function planBalanceDrawdown(fundingPlan, sourcesPreference)
+	local availableBalance = balances.getBalance(fundingPlan.address)
+	if sourcesPreference == "balance" or sourcesPreference == "any" then
+		fundingPlan.balance = math.min(availableBalance, fundingPlan.shortfall)
+		fundingPlan.shortfall = fundingPlan.shortfall - fundingPlan.balance
+	end
+end
```
Comment (dtfiedler) [inline src/gar.lua:1240]: note: bc these are not exported you likely cannot unit test

File: src/gar.lua [hunk 3]
```diff
@@ -1191,4 +1196,259 @@ function gar.delegateAllowedToStake(delegateAddress, gateway)
...
+			delegatedStake = stakeToDraw,
+			vaults = {}, -- set up vault spend tracking now while we're passing through
+		}
+		fundingPlan.shortfall = fundingPlan.shortfall - stakeToDraw
+		nextGateway.delegate.delegatedStake = nextGateway.delegate.delegatedStake - stakeToDraw
+		-- maintain consistency
+		nextGateway.delegate.excessStake = excessStake - stakeToDraw
+		nextGateway.totalDelegatedStake = nextGateway.totalDelegatedStake - stakeToDraw
+		gatewayIndex, nextGateway = next(gatewaysInfo, gatewayIndex)
+	end
```
Comment (dtfiedler) [inline src/gar.lua:1303]: for readability's sake you could also use for loop to avoid abstract use of `next` and potential infinite while loop

File: src/gar.lua [hunk 4]
```diff
@@ -1191,4 +1196,259 @@ function gar.delegateAllowedToStake(delegateAddress, gateway)
...
+		local stakeToDraw = math.min(excessStake, fundingPlan.shortfall)
+		fundingPlan["stakes"][nextGateway.delegate.gatewayAddress] = {
+			delegatedStake = stakeToDraw,
+			vaults = {}, -- set up vault spend tracking now while we're passing through
+		}
+		fundingPlan.shortfall = fundingPlan.shortfall - stakeToDraw
+		nextGateway.delegate.delegatedStake = nextGateway.delegate.delegatedStake - stakeToDraw
+		-- maintain consistency
+		nextGateway.delegate.excessStake = excessStake - stakeToDraw
+		nextGateway.totalDelegatedStake = nextGateway.totalDelegatedStake - stakeToDraw
```
Comment (dtfiedler) [inline src/gar.lua:1301]: can you explain why this? are you updating the reference to the current gateway to avoid double counting the stake available from it? more detailed comment might be helpful

File: src/gar.lua [hunk 5]
```diff
@@ -1191,4 +1196,259 @@ function gar.delegateAllowedToStake(delegateAddress, gateway)
...
+			end
+			return acc
+		end, {}),
+		{
+			{
+				order = "asc",
+				field = "endTimestamp",
+			},
+		}
+	)
```
Comment (dtfiedler) [inline src/gar.lua:1327]: the nested sort/reduce are hard to follow. may be helpful to breakup, at the sacrifice of memory footprint

File: src/gar.lua (hunk 2 above)
Comment (arielmelendez) [reply src/gar.lua:1240]: Correct - the are tested in integration with the top level function.

File: src/gar.lua (hunk 3 above)
Comment (arielmelendez) [reply src/gar.lua:1303]: Dang for some reason I convinced myself that you can't break out of a for loop in lua 🤔 But you can. I will make that adjustment.

File: src/gar.lua (hunk 4 above)
Comment (dtfiedler) [reply src/gar.lua:1301]: ahhh i see - makes sense

File: src/gar.lua (hunk 3 above)
Comment (arielmelendez) [reply src/gar.lua]: 4d392ad

File: src/gar.lua (hunk 4 above)
Comment (arielmelendez) [reply src/gar.lua]: Expanded comment:
98f6b38

File: src/gar.lua (hunk 5 above)
Comment (arielmelendez) [reply src/gar.lua]: b7f254d
